src/boot/efi/secure-boot.c

```diff
@@ -11,3 +14,121 @@ BOOLEAN secure_boot_enabled(void) {
+        EFI_STATUS err;
+
+        setup = audit = deployed = FALSE;
+
+        err = efivar_get_boolean_u8(EFI_GLOBAL_GUID, L"SetupMode", &setup);
+        if (EFI_ERROR(err))
+                return err;
+
+        err = efivar_get_boolean_u8(EFI_GLOBAL_GUID, L"AuditMode", &audit);
+        if (EFI_ERROR(err) && err != EFI_NOT_FOUND)
+                return err;
+
+        err = efivar_get_boolean_u8(EFI_GLOBAL_GUID, L"DeployedMode", &deployed);
+        if (EFI_ERROR(err) && err != EFI_NOT_FOUND)
+                return err;
+
+        *active = setup && !audit && !deployed;
+
+        return EFI_SUCCESS;
+}
+
+typedef enum {
+        SECURE_BOOT_KEY_PK,
+        SECURE_BOOT_KEY_KEK,
+        SECURE_BOOT_KEY_DB,
+        _SECURE_BOOT_KEY_TYPE_MAX,
+} SecureBootKeyType;
+
+static const CHAR16 *const key_type_string[_SECURE_BOOT_KEY_TYPE_MAX] = {
+        [SECURE_BOOT_KEY_PK] = L"PK",
+        [SECURE_BOOT_KEY_KEK] = L"KEK",
+        [SECURE_BOOT_KEY_DB] = L"db",
+};
+
+static const EFI_GUID *key_type_guid[_SECURE_BOOT_KEY_TYPE_MAX] = {
+        [SECURE_BOOT_KEY_PK] = EFI_GLOBAL_GUID,
+        [SECURE_BOOT_KEY_KEK] = EFI_GLOBAL_GUID,
+        [SECURE_BOOT_KEY_DB] = EFI_IMAGE_SECURITY_DATABASE_GUID,
+};
+
+EFI_STATUS setup_secure_boot(EFI_FILE *root_dir) {
+        BOOLEAN in_setup;
+        EFI_FILE_HANDLE secure_boot_dir;
+        EFI_STATUS err;
+
+        err = setup_mode_active(&in_setup);
+        if (EFI_ERROR(err)) {
+                Print(L"secure-boot: Failed to query setup mode state: %r", err);
+                return err;
+        }
+
+        if (!in_setup)
+                return EFI_SUCCESS;
+
+        err = uefi_call_wrapper(
+                root_dir->Open, 5, root_dir, &secure_boot_dir, L"\\loader\\secure-boot", EFI_FILE_MODE_READ, 0ULL);
+        if (err == EFI_NOT_FOUND)
+                return EFI_SUCCESS;
+        if (EFI_ERROR(err)) {
+                Print(L"secure-boot: Failed to open loader/secure-boot directory: %r", err);
+                return err;
+        }
+
+        for (;;) {
+                CHAR16 buf[256];
+                UINTN bufsize = sizeof(buf);
+                EFI_FILE_INFO *f;
+                _cleanup_freepool_ CHAR8 *content = NULL;
+                UINTN content_size;
+                SecureBootKeyType type;
+
+                err = uefi_call_wrapper(secure_boot_dir->Read, 3, secure_boot_dir, &bufsize, buf);
+                if (bufsize == 0 || EFI_ERROR(err))
+                        break;
+
+                f = (EFI_FILE_INFO *) buf;
+                if (f->FileName[0] == '.')
+                        continue;
+                if (f->Attribute & EFI_FILE_DIRECTORY)
+                        continue;
+
+                if (endswith(f->FileName, L".pk.esl"))
+                        type = SECURE_BOOT_KEY_PK;
+                else if (endswith(f->FileName, L".kek.esl"))
+                        type = SECURE_BOOT_KEY_KEK;
+                else if (endswith(f->FileName, L".db.esl"))
```

Comment:
hmm, FAT is traditionally case-insensitive but case-preserving... I figure we should probably do a case insensitive match here... (i wonder if we do this correctly everywhere else actually... we probably do not)